strategy_infura.go

```diff
@@ -0,0 +1,443 @@
+	// add a timeout
+	ctx, cancel := context.WithTimeout(ctx, time.Hour)
+	defer cancel()
+
+	sess := session.Must(session.NewSession(aws.NewConfig().WithRegion(region)))
+
+	start := time.Now()
+
+	originalAsg, err := getASGDetails(ctx, sess, asgName)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Refresh (strategy: infura-refresh) started on ASG %s\n", asgName)
+	fmt.Printf("Existing instances:\n")
+	for _, inst := range originalAsg.instances {
+		fmt.Printf("\t%s\n", inst)
+	}
+
+	// Record rollbacks during the procedure to get back on a reasonable state if the context get cancelled
+	// var rollbacks []func(ctx context.Context) error
+	//
+	// rollbacks = append(rollbacks, func(ctx context.Context) error {
+	// 	return setASGCapacities(ctx, as, asgName, originalAsg.maxSize, originalAsg.desiredCapacity)
+	// })
+
+	// Set instance protection on the old instances to be sure we keep having them around in case of failure
+	fmt.Printf("Enable instance protection on old instances\n")
+	err = setInstanceProtection(ctx, sess, asgName, originalAsg.instances)
+	if err != nil {
+		return err
+	}
+
+	// Double the ASG sizes
+	fmt.Printf("Double the ASG size:\n\tmaxSize %d-->%d\n\tdesiredCapacity %d-->%d\n",
+		originalAsg.maxSize, 2*originalAsg.maxSize, originalAsg.desiredCapacity, 2*originalAsg.desiredCapacity)
+	err = setASGCapacities(ctx, sess, asgName, 2*originalAsg.maxSize, 2*originalAsg.desiredCapacity)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Detect new instances and wait for them to be healthy\n")
+	newInstances, err := waitForNewInstances(ctx, sess, asgName, int(originalAsg.desiredCapacity), originalAsg.tg, originalAsg.instances)
+	if err != nil {
+		// rollback !
+		// TODO
+	}
+
+	// protect the new instances
+	fmt.Printf("Enable instance protection on the new instances\n")
+	err = setInstanceProtection(ctx, sess, asgName, newInstances)
+	if err != nil {
+		// rollback !
+		// TODO
+	}
+
+	// unprotect the old instances
+	fmt.Printf("Remove instance protection on the old instances\n")
+	err = removeInstanceProtection(ctx, sess, asgName, originalAsg.instances)
+	if err != nil {
+		// rollback !
+		// TODO
+	}
+
+	// get back to normal size, cull the old instances
+	fmt.Printf("Set back the ASG size:\n\tmaxSize %d-->%d\n\tdesiredCapacity %d-->%d\n",
+		2*originalAsg.maxSize, originalAsg.maxSize, 2*originalAsg.desiredCapacity, originalAsg.desiredCapacity)
+	err = setASGCapacities(ctx, sess, asgName, originalAsg.maxSize, originalAsg.desiredCapacity)
+	if err != nil {
+		// rollback !
+		// TODO
+		return err
+	}
+
+	fmt.Printf("Wait for scaling in\n")
+	err = waitForInstanceCount(ctx, sess, asgName, int(originalAsg.desiredCapacity))
+	if err != nil {
+		// rollback !
+		// TODO
+		return err
+	}
+
+	// unprotect the new instances
+	fmt.Printf("Remove instance protection on the new instances\n")
+	err = removeInstanceProtection(ctx, sess, asgName, newInstances)
```

Comment:
Any reason we want to not just leave the scale in protection in place permanently?

Reply:
To not assume anything about the ASG. There might be autoscaling rules there so leaving the protection might mess with that.

Reply:
Fair to support the general case.